evm/src/permutation.rs

```diff
@@ -68,6 +68,32 @@ impl<F: Field> GrandProductChallenge<F> {
     }
 }
 
+impl GrandProductChallenge<Target> {
+    pub(crate) fn combine_circuit<F: RichField + Extendable<D>, const D: usize>(
+        &self,
+        builder: &mut CircuitBuilder<F, D>,
+        terms: &[ExtensionTarget<D>],
+    ) -> ExtensionTarget<D> {
+        let zero = builder.zero();
+        let mut factor = ReducingFactorTarget::new(self.beta.to_ext_target(zero));
+        let reduced = factor.reduce(terms, builder);
+        builder.add_extension(reduced, self.gamma.to_ext_target(zero))
+    }
+
+    pub(crate) fn combine_base_circuit<F: RichField + Extendable<D>, const D: usize>(
+        &self,
+        builder: &mut CircuitBuilder<F, D>,
+        terms: &[Target],
+    ) -> Target {
+        let zero = builder.zero();
+        let combination = terms
```

Comment:
Could use `reduce_base`, to save a little code and potentially use `ReducingGate` in some cases

Reply:
Hm so `reduce_base` takes a reducing factor that is an `ExtensionTarget`, whereas here the factor `self.beta` is just a `Target`. `reduce_base` will either use `ArithmeticExtensionGate` operations or add a new `ReducingGate`.
I think if the number of terms is low (like `< num_routed_wires/4`), using `ArithmeticGate` operation is cheaper. I think I'll add a util function for this kind of reduction.